pilot/pkg/config/kube/gateway/deploymentcontroller.go

```diff
@@ -532,7 +620,144 @@ func (d *DeploymentController) render(templateName string, mi TemplateInput) ([]
 		return nil, err
 	}
 
-	return yml.SplitString(results), nil
+	rawOutput := yml.SplitString(results)
+	transformedOutput := make([]string, 0, len(rawOutput))
+	for _, output := range rawOutput {
+		to, err := applyOverlay(output, templateOverlays)
+		if err != nil {
+			return nil, err
+		}
+		if to != "" {
+			transformedOutput = append(transformedOutput, to)
+		}
+	}
+	return transformedOutput, nil
+}
+
+func oldestConfig[T controllers.Object](configs []T) T {
+	return slices.MinFunc(configs, func(i, j T) int {
+		if r := i.GetCreationTimestamp().Compare(j.GetCreationTimestamp().Time); r != 0 {
+			return r
+		}
+		// If creation time is the same, then behavior is nondeterministic. In this case, we can
+		// pick an arbitrary but consistent ordering based on name and namespace, which is unique.
+		// CreationTimestamp is stored in seconds, so this is not uncommon.
+		if r := cmp.Compare(i.GetName(), j.GetName()); r != 0 {
+			return r
+		}
+		return cmp.Compare(i.GetNamespace(), j.GetNamespace())
+	})
+}
+
+var supportedOverlays = sets.New(
+	"deployment",
+	"service",
+	"serviceAccount",
+	"horizontalPodAutoscaler",
+	"podDisruptionBudget",
+)
+
+var requiredOverlays = sets.New(
+	"horizontalPodAutoscaler",
+	"podDisruptionBudget",
+)
+
+func applyOverlay(object string, overlaysList []map[string]string) (string, error) {
+	var ik crd.IstioKind
+	if err := yaml.Unmarshal([]byte(object), &ik); err != nil {
+		return "", fmt.Errorf("failed to find kind: %v", err)
+	}
+	gv, err := schema.ParseGroupVersion(ik.TypeMeta.APIVersion)
+	if err != nil {
+		return "", fmt.Errorf("failed to find kind: %v", err)
+	}
+	kind := &schema.GroupVersionKind{Group: gv.Group, Version: gv.Version, Kind: ik.TypeMeta.Kind}
+
+	var data any
+	var key string
+	switch kind.Kind {
+	case gvk.Deployment.Kind:
+		data = &appsv1.Deployment{}
+		key = "deployment"
+	case gvk.Service.Kind:
+		data = &corev1.Service{}
+		key = "service"
+	case gvk.ServiceAccount.Kind:
+		data = &corev1.ServiceAccount{}
+		key = "serviceAccount"
+	case gvk.HorizontalPodAutoscaler.Kind:
+		data = &autoscalingv2.HorizontalPodAutoscaler{}
+		key = "horizontalPodAutoscaler"
+	case gvk.PodDisruptionBudget.Kind:
+		data = &policyv1.PodDisruptionBudget{}
+		key = "podDisruptionBudget"
+	default:
+		return "", fmt.Errorf("unknown overlay kind %q", kind.Kind)
+	}
+	applied := false
+	for _, overlays := range overlaysList {
+		for k := range overlays {
+			if !supportedOverlays.Contains(k) {
+				return "", fmt.Errorf("unsupported overlay %q (supported: %v)", k, sets.SortedList(supportedOverlays))
```

Comment:
Should we really not apply any overlay if there's an invalid one in there? Probably safer, but does cause more friction

Reply:
this doesn't just block the overlay, it blocks any writing at all. I think this is correct; imagine my overlay added critical security functionality ("do not run as root"), we don't want to fail open